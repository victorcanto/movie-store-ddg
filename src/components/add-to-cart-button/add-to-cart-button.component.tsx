import React, { useEffect, useState } from 'react';
import { Typography, useTheme } from '@mui/material';

import {
	addCartItem,
	checkCartItemExistence,
	FavoriteItem,
	Movie,
	removeFavoriteItem,
} from '../../shared';
import {
	selectCartItems,
	setCartItems,
	useAppDispatch,
	useAppSelector,
} from '../../store';
import { Button } from '../button/button.component';

export interface AddToCartButtonProps {
	item: Movie;
}

export function AddToCartButton({ item }: AddToCartButtonProps) {
	const theme = useTheme();
	const dispatch = useAppDispatch();
	const cartItems = useAppSelector(selectCartItems);
	const [isAddedToCart, setIsAddedToCart] = useState<boolean>(false);

	useEffect(() => {
		const match = checkCartItemExistence(cartItems, item);

		if (match) {
			setIsAddedToCart(true);
		} else {
			setIsAddedToCart(false);
		}
	}, [cartItems]);

	const toggle = (): void => setIsAddedToCart(!isAddedToCart);

	const addItemToCartHandler = (): void => {
		const newCartItems = addCartItem(cartItems, item);

		if (newCartItems) {
			dispatch(setCartItems(newCartItems));
		}
	};

	const removeItemToCartHandler = (): void => {
		const newCartItems = removeFavoriteItem(cartItems, item);

		if (newCartItems) {
			dispatch(setCartItems(newCartItems));
		}
	};

	useEffect(() => {
		if (isAddedToCart) {
			addItemToCartHandler();
		} else {
			removeItemToCartHandler();
		}
	}, [isAddedToCart]);

	return (
		<Button
			data-testid='add-to-cart-button'
			onClick={toggle}
			sx={{
				p: '1rem',
			}}
		>
			<Typography
				noWrap
				variant='h6'
				fontWeight={600}
				textAlign='center'
				color={theme.palette.common.white}
			>
				{isAddedToCart ? 'REMOVER' : 'ADICIONAR'}
			</Typography>
		</Button>
	);
}
