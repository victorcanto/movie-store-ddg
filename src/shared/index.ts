export * from './models/cart.interfaces';
export * from './models/favorites.interfaces';
export * from './models/movies.interfaces';
export * from './utils/store/cart.utils';
export * from './utils/store/favorites.utils';
export * from './utils/test-utils';
export * from './utils/currency.utils';
export * from './utils/states-and-cities';
