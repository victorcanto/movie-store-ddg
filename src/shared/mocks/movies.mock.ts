import { Movie } from '../models/movies.interfaces';

export const moviesMock: Movie[] = [
	{
		adult: false,
		backdrop_path: '/ugS5FVfCI3RV0ZwZtBV3HAV75OX.jpg',
		genre_ids: [16, 878, 28],
		genres: [
			{ id: 16, name: 'Animação' },
			{ id: 878, name: 'Ficção científica' },
			{ id: 28, name: 'Ação' },
		],
		id: 610150,
		original_language: 'ja',
		original_title: 'ドラゴンボール超 スーパーヒーロー',
		overview:
			'The Red Ribbon Army, an evil organization that was once destroyed by Goku in the past, has been reformed by a group of people who have created new and mightier Androids, Gamma 1 and Gamma 2, and seek vengeance against Goku and his family.',
		popularity: 16585.915,
		poster_path: '/rugyJdeoJm7cSJL1q4jBpTNbxyU.jpg',
		release_date: '2022-06-11',
		title: 'Dragon Ball Super: Super Hero',
		video: false,
		vote_average: 8,
		vote_count: 959,
		price: 79.99,
	},
	{
		adult: false,
		backdrop_path: '/vvObT0eIWGlArLQx3K5wZ0uT812.jpg',
		genre_ids: [28, 12, 14],
		genres: [
			{ id: 16, name: 'Animação' },
			{ id: 878, name: 'Ficção científica' },
			{ id: 28, name: 'Ação' },
		],
		id: 616037,
		original_language: 'en',
		original_title: 'Thor: Love and Thunder',
		overview:
			'After his retirement is interrupted by Gorr the God Butcher, a galactic killer who seeks the extinction of the gods, Thor Odinson enlists the help of King Valkyrie, Korg, and ex-girlfriend Jane Foster, who now inexplicably wields Mjolnir as the Relatively Mighty Girl Thor. Together they embark upon a harrowing cosmic adventure to uncover the mystery of the God Butcher’s vengeance and stop him before it’s too late.',
		popularity: 5372.967,
		poster_path: '/pIkRyD18kl4FhoCNQuWxWu5cBLM.jpg',
		release_date: '2022-07-06',
		title: 'Thor: Love and Thunder',
		video: false,
		vote_average: 6.8,
		vote_count: 2223,
		price: 79.99,
	},
	{
		adult: false,
		backdrop_path: '/xfNHRI2f5kHGvogxLd0C5sB90L7.jpg',
		genre_ids: [16, 28, 10751, 878, 35],
		genres: [
			{ id: 16, name: 'Animação' },
			{ id: 878, name: 'Ficção científica' },
			{ id: 28, name: 'Ação' },
			{
				id: 10751,
				name: 'Família',
			},
			{
				id: 878,
				name: 'Ficção científica',
			},
			{
				id: 35,
				name: 'Comédia',
			},
		],
		id: 539681,
		original_language: 'en',
		original_title: 'DC League of Super-Pets',
		overview:
			'When Superman and the rest of the Justice League are kidnapped, Krypto the Super-Dog must convince a rag-tag shelter pack - Ace the hound, PB the potbellied pig, Merton the turtle and Chip the squirrel - to master their own newfound powers and help him rescue the superheroes.',
		popularity: 4630.633,
		poster_path: '/r7XifzvtezNt31ypvsmb6Oqxw49.jpg',
		release_date: '2022-07-27',
		title: 'DC League of Super-Pets',
		video: false,
		vote_average: 7.6,
		vote_count: 360,
		price: 79.99,
	},
];
