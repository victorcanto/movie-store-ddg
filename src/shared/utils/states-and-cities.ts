import statesAndCities from '../data/estados-cidades.json';

export const getBrazilianStates = () =>
	statesAndCities.estados.map(({ nome }) => nome);

export const getBrazilianCitiesByState = (stateName: string) =>
	statesAndCities.estados.find(({ nome }) => nome === stateName)?.cidades;
