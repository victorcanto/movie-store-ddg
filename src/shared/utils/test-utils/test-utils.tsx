import React, { PropsWithChildren } from 'react';
import { render as rtlRender } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { setupStore, AppStore, RootState } from '../../../store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}

export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {},
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return (
			<Provider store={store}>
				<BrowserRouter>{children}</BrowserRouter>
			</Provider>
		);
	}
	return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export { renderWithProviders as render };
