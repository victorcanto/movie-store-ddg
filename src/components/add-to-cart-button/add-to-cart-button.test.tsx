import React from 'react';
import { moviesMock } from '../../shared/mocks';
import { render } from '../../shared/utils/test-utils';
import { AddToCartButton } from './add-to-cart-button.component';

test('snapshot of the base state', () => {
	const { asFragment } = render(<AddToCartButton item={moviesMock[0]} />);
	expect(asFragment()).toMatchSnapshot();
});

test('should have in the document', () => {
	const { getByTestId } = render(<AddToCartButton item={moviesMock[0]} />);
	expect(getByTestId('add-to-cart-button')).toBeInTheDocument();
});
