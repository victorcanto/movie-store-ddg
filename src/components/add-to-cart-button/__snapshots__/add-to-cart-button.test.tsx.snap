// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-rf0srl-MuiButtonBase-root-MuiButton-root"
    data-testid="add-to-cart-button"
    tabindex="0"
    to="/"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-b2zrw0-MuiTypography-root"
    >
      ADICIONAR
    </h6>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
